Feature / update the prplMesh utils operation test to use BML

itayx opened the pull request:
prplmesh_utils.sh status returns an operational status determined by parsing the logs.
In some platforms logs wrap around could return "fail" status although system is operational state.
Currently, prplmesh_utils.sh status returns an operational status determined by parsing the logs.

To fix this, the prplmesh_utils can use the BML command `bml_get_device_operational_radios`

kantera800 reviewed (commented):
You should mention in the 2nd commit message that after this change device without a controller process running on it  (repeater for example ) will not support `prplmesh_utils.sh status`

Also, you need a commit to update test_status.sh  so it won't run on repeater or at least not to expect an answer. I prefer the first option.

vitalybu commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -246,27 +246,20 @@ stop_function() {
-        success "OK $msg"
+execute_bml() {
+    error=0
+    bml_cli="$PRPLMESH_BIN_DIR/beerocks_cli"
```
Why? Just use `beerocks_cli` directly. No need for additional aliases.

itayx replied:
Removed the unnecessary aliases and global variables

vitalybu commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -246,27 +246,20 @@ stop_function() {
-    msg="$1"; shift
-    if "$@"; then
-        success "OK $msg"
+execute_bml() {
```
It's basically a wrapper for `beerocks_cli`, so not sure why it's called `execute_bml`.

itayx replied:
changed the name

vitalybu commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -246,27 +246,20 @@ stop_function() {
+    bml_cli="$PRPLMESH_BIN_DIR/beerocks_cli"
+    tmp_file=$(mktemp)
+    if [ -e $bml_cli ]; then
+        $bml_cli -c "$@" > $tmp_file &
```
What's the point of running the command in the background and then waiting for it to finish?
Why not do something like this: `output=$(${PRPLMESH_BIN_DIR}/beerocks_cli -c "$@")`?

itayx replied:
When I tried it the first time i must have wrote it wrong,
Changed and removed the unnecessary commands (the use of tmp_file etc.)

vitalybu commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -276,21 +269,22 @@ status_function() {
 
-    # check for operational status
-    LOGS_PATH=@BEEROCKS_LOG_FILES_PATH@
+    bridge_mac="$(ifconfig @BEEROCKS_BRIDGE_IFACE@ | grep -o -E '([[:xdigit:]]{1,2}:){5}[[:xdigit:]]{1,2}')"
```
In dockers we use a slightly different command for extracting the bridge mac address:
`bridge_mac="$(ip link show dev br-lan | awk '/^ *link\/ether / {print $2}')"`

To me it looks a bit less complex.

itayx replied:
You're right
Changed

vitalybu commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -276,21 +269,22 @@ status_function() {
-    LOGS_PATH=@BEEROCKS_LOG_FILES_PATH@
+    bridge_mac="$(ifconfig @BEEROCKS_BRIDGE_IFACE@ | grep -o -E '([[:xdigit:]]{1,2}:){5}[[:xdigit:]]{1,2}')"
+    bml_cmd="bml_get_device_operational_radios $bridge_mac"
+    dbg "Executing BML: $bml_cmd"
```
You're executing CLI commands, not BML

itayx replied:
Renamed

vitalybu commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -276,21 +269,22 @@ status_function() {
+    bml_cmd="bml_get_device_operational_radios $bridge_mac"
+    dbg "Executing BML: $bml_cmd"
 
     error=0
```
I don't like the use of global variables.
You could, for example, return a string from this function, and decide that an empty string means an error occurred.

itayx replied:
Removed the unnecessary aliases and global variables

vitalybu commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -276,21 +269,22 @@ status_function() {
-    exit $error
+    execute_bml "$bml_cmd"
+    dbg $output
+    OK_count=`echo "$output" | grep -c -e "OK.*operational"`
```
You should probably add a comment describing the expected output and what exactly this conditions means

itayx replied:
Added a comment of the expected output

itayx committed:
common: prplmesh_utils.sh: add bml execute func

This is a preparative commit for "common: prplmesh_utils.sh: use BML for
operational check"

Add execute BML function

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

itayx committed:
common: prplmesh_utils.sh: use BML operational check

prplmesh_utils.sh status returns an operational status determined by
parsing the logs. In some platforms logs wrap around could return "fail"
status although system is operational state.

To resolve this issue, the status check should use the BML command
bml_get_device_operational_radios which returns the status of the
current device using the bridge's MAC address.

Use BML to check operational status

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

yaniv-yitzhak commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -287,21 +287,46 @@ status_function() {
 
-    # check for operational status
-    LOGS_PATH=@BEEROCKS_LOG_FILES_PATH@
+    bridge_mac="$(ip link show dev br-lan | awk '/^ *link\/ether / {print $2}')"
```
Bridge name is not always br-lan, it can be brlan0 or anything else

itayx replied:
This is why it fails on RDKB

The bridge name needs to be fetched with `@BEEROCKS_BRIDGE_IFACE@`

itayx replied:
Fixed

yaniv-yitzhak reviewed (changes requested):
Need generic way to get the bridge name or interface names

vitalybu commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -246,6 +246,17 @@ stop_function() {
+    dbg "Executing beerocks cli command: $*"
+    if [ -e $PRPLMESH_BIN_DIR/beerocks_cli ]; then
+        output=$(${PRPLMESH_BIN_DIR}/beerocks_cli -c "$@")
+        return $?
```
Why do you return the exit code of the operation instead of the actual output?

itayx replied:
The output is stored in the global `output` variable
while the exit code is returned
Isn't that proper code functionality?

vitalybu replied:
Well, you could return just an empty string in case of a failure..
You don't use the exit code anyway

itayx replied:
There is a validation of the bml exit code `if execute_beerocks_command "$bml_cmd" ; then`
If the beerocks_cli doesn't exist it would then return an error code of 1

vitalybu commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -287,21 +287,46 @@ status_function() {
+    bridge_mac="$(ip link show dev @BEEROCKS_BRIDGE_IFACE@ | awk '/^ *link\/ether / {print $2}')"
+    bml_cmd="bml_get_device_operational_radios $bridge_mac"
+
+   
```
Remove extra new-line

itayx replied:
removed

vitalybu commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -287,21 +287,46 @@ status_function() {
-    # check for operational status
-    LOGS_PATH=@BEEROCKS_LOG_FILES_PATH@
+    bridge_mac="$(ip link show dev @BEEROCKS_BRIDGE_IFACE@ | awk '/^ *link\/ether / {print $2}')"
+    bml_cmd="bml_get_device_operational_radios $bridge_mac"
```
It's used only once, why do you need a variable for that?

itayx replied:
cleaner code.
When I write it in a single line
It's harder to understand, this way it's clear that the BML command uses the bridge's mac address for it's functionality

itayx replied:
```suggestion
    bml_cmd="bml_get_device_operational_radios $(ip link show dev @BEEROCKS_BRIDGE_IFACE@ | awk '/^ *link\/ether / {print $2}')
```

This line is harder to understand then using a variable for the bridge's mac

vitalybu replied:
Ok

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -246,6 +246,17 @@ stop_function() {
+        output=$(${PRPLMESH_BIN_DIR}/beerocks_cli -c "$@")
+        return $?
+    else
+        err "BML CLI not found"
```
You should set `output=''` here.

itayx replied:
Added

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -287,21 +287,45 @@ status_function() {
+    bridge_mac="$(ip link show dev @BEEROCKS_BRIDGE_IFACE@ | awk '/^ *link\/ether / {print $2}')"
+    bml_cmd="bml_get_device_operational_radios $bridge_mac"
+
+    if  pgrep -l beerocks | grep -q beerocks_cont ; then
```
Can't we just do
```suggestion
    if  pgrep -l beerocks_controller
```

itayx replied:
For some reason the pgrep output is cut short and cannot find "beerocks_controller"
So I changed it to use
 ` if  pgrep -l beerocks_cont ; then` instead

arnout reviewed (approved):
Nitpicks only.

rmelotte commented:
@vitalybu Your review status is still "change requested" so it won't be merged automatically.
Can you approve it (or re-review it)?

rmelotte commented:
@mergifyio rebase
